### PR TITLE
feat(iroh-net): PCP probe

### PR DESCRIPTION
**Pull request by divagant-martian**
## Description
- Adds PCP support for the `Announce Opcode` and implemented probing for PCP
- Add metrics for PCP probes
- Add a `port-map-probe` command to pick what subset of protocols to probe

## Notes & open questions
- This is the pcp probes stripped down from #1209, mappings take the larger part of the code and I'll add them in a next PR to keep the changes short,the other one was getting extremely big
- Some structures like the `OpcodeData` might look redundant right not but allow preparation for pcp mappings

Suggested way of trying it:
```bash
env RUST_LOG=iroh_net::hp::portmapper=trace cargo run doctor port-map-probe --enable-pcp --enable-upnp
```
local results:
```log
probing port mapping protocols with Config { enable_upnp: true, enable_pcp: true, enable_nat_pmp: false }
DEBUG portmapper.service: iroh_net::hp::portmapper: portmap starting
TRACE portmapper.service: iroh_net::hp::portmapper: tick: msg Some(Probe { result_tx: _ })
TRACE portmapper.service:portmapper.probe: iroh_net::hp::portmapper::pcp: probe response: Response { lifetime_seconds: 0, epoch_time: 1203479, data: Announce }
TRACE portmapper.service:portmapper.probe: iroh_net::hp::portmapper: tick: pcp probe ready
TRACE portmapper.service:portmapper.probe: iroh_net::hp::portmapper: tick: upnp probe ready
TRACE portmapper.service: iroh_net::hp::portmapper: tick: probe ready
DEBUG portmapper.service: iroh_net::hp::portmapper: upnp gateway changed "None" -> "http://192.168.20.1:58089/ctl/IPConn"
DEBUG portmapper.service: iroh_net::hp::portmapper: probe output portmap={ UPnP: true, PMP: false, PCP: true }
portmap={ UPnP: true, PMP: false, PCP: true }
```

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -3642,6 +3671,28 @@ version = "0.1.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "e9426b2a0c03e6cc2ea8dbc0168dbbf943f88755e409fb91bcb8f6a268305f4a"
 
+[[package]]
```
Comment (dignifiedquire) [inline Cargo.lock:3674]: looks like you need to run `cargo update` to get rid of this unused dep

File: iroh-metrics/src/portmap.rs [hunk 2]
```diff
@@ -44,6 +50,12 @@ impl Default for Metrics {
             upnp_gateway_updated: Counter::new(
                 "Number of UPnP probes that resulted in a gateway different to the previous one.",
             ),
+
+            /*
+             * PCP metrics
+             */
+            pcp_probes: Counter::new("Number of PCP probes executed."),
+            pcp_available: Counter::new("Number of PCP probes that found it available."),
```
Comment (dignifiedquire) [inline iroh-metrics/src/portmap.rs:58]: who or what is `it` here?

File: iroh/src/commands/doctor.rs [hunk 3]
```diff
@@ -102,6 +102,15 @@ pub enum Commands {
...
         local_derper: bool,
     },
+    /// Probe the port mapping protocols.
+    PortMapProbe {
+        /// Whether to enable UPnP.
+        #[clap(long)]
+        enable_upnp: bool,
+        /// Whether to enable PCP.
+        #[clap(long)]
+        enable_pcp: bool,
```
Comment (dignifiedquire) [inline iroh/src/commands/doctor.rs:112]: maybe make both default to `true`?

File: iroh-metrics/src/portmap.rs (hunk 2 above)
Comment (divagant-martian) [reply iroh-metrics/src/portmap.rs:58]: PCP? how would you phrase it?

File: Cargo.lock (hunk 1 above)
Comment (divagant-martian) [reply Cargo.lock]: [549075e](https://github.com/n0-computer/iroh/pull/1241/commits/549075e5a679827246f00a3b24a45fab16098448) should do it

File: iroh/src/commands/doctor.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh/src/commands/doctor.rs:125]: ah right, I remember

File: iroh-metrics/src/portmap.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-metrics/src/portmap.rs:58]: phrasing is fine, I was just having some trouble understanding "it" ;)

File: iroh-net/src/hp/portmapper/pcp/protocol/response.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        1 + // result code
+        4 + // lifetime
+        4 + // epoch time
+        12; // reserved
+
+    /// Indicator ORd into the [`Opcode`] to indicate a response packet.
+    pub const RESPONSE_INDICATOR: u8 = 1u8 << 7;
+
+    /// Decode a response.
+    pub fn decode(buf: &[u8]) -> Result<Self, Error> {
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-net/src/hp/portmapper/pcp/protocol/response.rs:155]: and encoding

**Comment by dignifiedquire** [comment]
Tested the doctor command 

```
❯ iroh doctor port-map-probe --enable-pcp --enable-upnp
probing port mapping protocols with Config { enable_upnp: true, enable_pcp: true, enable_nat_pmp: false }
portmap={ UPnP: true, PMP: false, PCP: true }
```

File: iroh-net/src/hp/portmapper/pcp/protocol/response.rs [hunk 5]
```diff
@@ -197,4 +211,59 @@ impl Response {
             data,
         })
     }
+
+    #[cfg(test)]
+    fn random(opcode: Opcode) -> Self {
+        let data = OpcodeData::random(opcode);
+        Self {
+            lifetime_seconds: rand::random(),
```
Comment (dignifiedquire) [inline iroh-net/src/hp/portmapper/pcp/protocol/response.rs:219]: nit: I prefer using a passed in `Rng` for these, so you can use seeded randomness in tests (which is great when you have to debug these)

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**